mysql/provider.go

```diff
@@ -175,6 +194,41 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}
 	var allowNativePasswords = authPlugin == nativePasswords
 	var password = d.Get("password").(string)
 	var iam_auth = d.Get("iam_database_authentication").(bool)
+	var caCertPath = d.Get("ca_cert").(string)
+	var clientCertPath = d.Get("client_cert").(string)
+	var clientKeyPath = d.Get("client_key").(string)
+	var tlsConfig *tls.Config
+
+	certSettings := lo.Filter([]string{caCertPath, clientCertPath, clientKeyPath}, func(x string, index int) bool {
```

Comment:
Just nitpicking, this is not necessary: consider using nested schema like `custom_tls`, that would have required 3 arguments. Wouldn't that work and be more user friendly?

Reply:
Sure thing. Go is definitely not a language I'm very familiar with, so please correct me on style :)

Reply:
Generally, Go is a bit more talkative and I don't like functions like this that much.

That said, I would accept that - but terraform has nested schemas and I wondered, whether they cannot be utilized. (Just to be sure: I am not 100% convinced that it will work, but it could).

Reply:
I played around a bit with this and I think introducing nested attributes might be a considerable amount of changes that might make it a bit messy. Happy to discuss that approach, but I've moved it into a map/struct to make it a bit more understandable. I hope that's alright

Reply:
My idea was like this
```
diff --git a/mysql/provider.go b/mysql/provider.go
index da6ee51..77698ad 100644
--- a/mysql/provider.go
+++ b/mysql/provider.go
@@ -123,9 +123,30 @@ func Provider() *schema.Provider {
                        },
 
                        "custom_tls": {
-                               Type:     schema.TypeMap,
+                               Type:     schema.TypeList,
                                Optional: true,
-                               Default:  nil,
+                               MaxItems: 1,
+                               Elem: &schema.Resource{
+                                       Schema: map[string]*schema.Schema{
+                                               "config_key": {
+                                                       Type:     schema.TypeString,
+                                                       Default:  "custom",
+                                                       Optional: true,
+                                               },
+                                               "ca_cert": {
+                                                       Type:     schema.TypeString,
+                                                       Required: true,
+                                               },
+                                               "client_cert": {
+                                                       Type:     schema.TypeString,
+                                                       Required: true,
+                                               },
+                                               "client_key": {
+                                                       Type:     schema.TypeString,
+                                                       Required: true,
+                                               },
+                                       },
+                               },
                        },
 
                        "max_conn_lifetime_sec": {
@@ -193,13 +214,14 @@ func providerConfigure(ctx context.Context, d *schema.ResourceData) (interface{}
        var password = d.Get("password").(string)
        var iam_auth = d.Get("iam_database_authentication").(bool)
        var tlsConfig = d.Get("tls").(string)
-       var customTLSMap = make(map[string]interface{})
        var tlsConfigStruct *tls.Config
        var customTLS CustomTLS
 
-       customTLSMap = d.Get("custom_tls").(map[string]interface{})
+       customTLSMap := d.Get("custom_tls").([]interface{})
        if len(customTLSMap) > 0 {
-               customTLSJson, err := json.Marshal(customTLSMap)
+               // There is max 1 element based on schema, get it.
+               customMap := customTLSMap[0].(map[string]interface{})
+               customTLSJson, err := json.Marshal(customMap)
                if err != nil {
                        return nil, diag.Errorf("failed to marshal tls config: %v", customTLSMap)
                }
```

I tried it and it seems to work with config like this and it enforces having all elements specified and it could work with IDE autocompletion. I did not try the values there, but that is not so complicated.
```
provider "mysql" {
  endpoint = "127.0.0.1:3380"
  username = "root"
  password = "my-secret-pw"
  custom_tls {
	ca_cert = "a.crt"
	client_key = "cl.key"
	client_cert = "cl.crt"
  }
}
```

Reply:
Brilliant! That's much nicer, indeed. Many thanks!